sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala

```diff
@@ -176,15 +162,18 @@ object BindParameters extends ParameterizedQueryProcessor with QueryErrorsBase {
     }
   }
 
-  private def bind(p: LogicalPlan)(f: PartialFunction[Expression, Expression]): LogicalPlan = {
-    p.resolveExpressionsWithPruning(_.containsPattern(PARAMETER)) (f orElse {
-      case sub: SubqueryExpression => sub.withNewPlan(bind(sub.plan)(f))
-    })
+  private def bind(p0: LogicalPlan)(f: PartialFunction[Expression, Expression]): LogicalPlan = {
+    var stop = false
+    p0.resolveOperatorsDownWithPruning(_.containsPattern(PARAMETER) && !stop) {
+      case p1 =>
+        stop = p1.isInstanceOf[ParameterizedQuery]
```

Comment:
This looks a bit hacky. Shall we do a manual tree traversal so that we can stop recursion when we hit `ParameterizedQuery`?

Reply:
we can leave it and clean it up after TreeNode transform down methods supporting early stop

Reply:
@cloud-fan How about this API:
```scala
    p0.resolveOperatorsDownUntil(
      p1 => !p1.containsPattern(PARAMETER) || p1.isInstanceOf[ParameterizedQuery]) {
      f orElse {
        case sub: SubqueryExpression => sub.withNewPlan(bind(sub.plan)(f))
      }
    }
```
it executes body until `stopCondition: T => Boolean` returns `true` including the last operator. So, it is like `repeat .. until` in Pascal or `do .. while` in C/C++. In our case, we need to substitute parameters in variables of a `ParameterizedQuery` on which we have to stop.

Reply:
LGTM. So we don't want to use it together with the pruning condition?

Reply:
We can check the bits in `stopCondition`. Don't think we need a separate condition for pruning.